app/models/favorite_tag.rb

```diff
@@ -11,16 +11,24 @@ class FavoriteTag < ActiveRecord::Base
                                  }
 
   validate :within_limit, on: :create
-
-  after_save :expire_cached_home_favorite_tags
-  after_destroy :expire_cached_home_favorite_tags
-
   def within_limit
-    if user.favorite_tags(:reload).count >= ArchiveConfig.MAX_FAVORITE_TAGS
-      errors.add(:base, ts('Sorry, you can only save %{maximum} favorite tags.', maximum: ArchiveConfig.MAX_FAVORITE_TAGS))
+    if user && user.favorite_tags(:reload).count >= ArchiveConfig.MAX_FAVORITE_TAGS
+      errors.add(:base,
+                 ts("Sorry, you can only save %{maximum} favorite tags.",
+                    maximum: ArchiveConfig.MAX_FAVORITE_TAGS))
+    end
+  end
+
+  validate :canonical, on: :create
```

Comment:
What would happen if someone favorited a tag and it was changed to non canonical ?